src/HtmlReader/architecture.md

```diff
@@ -0,0 +1,78 @@
+- Inject any injectables. For example, we need to add the Readium-CSS stylesheets to the `<head>`.
+- Set initial CSS variables like scroll settings and color mode.
+- Add a simple JS script that listens for internal link clicks and calls `window.parent.postMessage` to send a message to the parent about the clicked link.
+
+Note that we _could_ do all of this in a `useEffect` hook after the resource loads, but then there would be a flash of unstyled content between the first render and when we add our Readium-CSS stylesheet. To avoid this, we simple add it to the document string before it ever gets to the `<iframe>` for rendering.
+
+## Updating the iframe
+
+It is sometimes necessary to reach in to the iframe and update it as the result of some state transition. To do this, we use `React.useEffect`. There are two primary times we need to reach in to the iframe and update it:
+
+1. When user settings change in the state, we need to update the CSS variables. This is a very simple hook in `useUpdateCSS`.
+2. When a user navigates within one resource, we need to manually scroll them to the correct location. For example, when a user clicks the "Next Page" button, we update the location state, which triggers a navigation effect. In this effect, we need to reach in to the iframe and set the `scrollLeft` or `scrollTop` values to the location of the next page, depending on if the user is in scrolling or paginated mode. (more below).
+
+### Inter-resource Navigation
+
+This might be the most challenging aspect of the web reader architecture. In the Readium Architecture, a `Locator` indicates a specific location. `locator.href` is the resource, and `locator.locations` is an object containing multiple references to the specific location within the resource. It can contain:
+
+- `progression`: A decimal value representing percent of the way through the resource. This is what we typically use when navigating to scroll positions. We calculate the `progression` of the desired position and scroll the user there.
+- `fragment`: A string representing a location. Currently, we only support anchors (eg: `"#some-heading"`). We can scroll the user there by querying the dom for that element and then calling `el.scrollIntoView()`.
+- `position`: An integer representation of the location, which I interpret to mean "page number". This is not very portable because the number of pages and size of pages changes when browser size changes or font size changes. Thus, we prefer to use `progression`, but we keep `position` updated in case we wish to display it at some point.
+
+Note on CFIs: In the future, we would like to support CFIs. These are strings representing specific locations in an HTML document. They are portable as well. When you're reading a kindle and it says "location 441/2387", those numbers are the CFI that you are on and the total number of CFIs. More work on this is needed in the future.
```

Comment:
What does CFI stand for?

Reply:
Canonical Fragment Identifier. I'm not totally sure how they work, but basically it's some string that points to a specific position in an HTML document.